utils/emmylua/wesnoth.lua

```diff
@@ -0,0 +1,272 @@
+---@meta
+
+wesnoth = {}
+
+---@return "'Game Lua Kernel'" | "'Mapgen Lua Kernel'" | "'Application Lua Kernel'"
+function wesnoth.kernel_type() end
+
+---@param path string
+---@return ...
+function wesnoth.dofile(path, ...) end
+
+---@class tstring : string
+
+---Constructs a textdomain, which can be called to create translatable strings.
+---@param domain string The textdomain name
+---@return fun(str:string, str_plural?:string, n?:number):tstring
+function wesnoth.textdomain(domain) end
+
+---Logs a message to the console
+---@param logger "'info'"|"'debug'"|"'warning'"|"'error'"|"'wml'"
+---@param message string
+---@param in_chat boolean
+---@overload fun(message:string, in_chat:boolean)
+function wesnoth.log(logger, message, in_chat) end
+
+---Simulate combat between two units_MP
+---@param attacker unit
+---@param attacker_weapon integer
+---@param defender unit
+---@param defender_weapon integer
+---@return stats_evaluation attacker_stats
+---@return stats_evaluation defender_stats
+---@return weapon_evaluation attacker_weapon
+---@return weapon_evaluation defender_weapon
+function wesnoth.simulate_combat(attacker, attacker_weapon, defender, defender_weapon) end
+
+---@class stats_evaluation
+---@field poisoned number The chance to be poisoned as a number in [0,1].
+---@field slowed number The chance to be slowed as a number in [0,1].
+---@field untouched number The chance to receive no damage as a number in [0,1].
+---@field average_hp number
+---@field hp_chance table<number, number> The chance to end the attack with a specific number of hit points. The key is the number of hit points, from 0 to max hitpoints. The value is the chance as a number in [0,1].
+
+---@class weapon_evaluation
+---@field num_blows integer The number of times the attack hits.
+---@field damage integer The damage the attack inflicts per hit.
+---@field chance_to_hit number The chance for the attack to hit.
+---@field poisons boolean Indicates whether the attack inflicts poison.
+---@field slows boolean Indicates whether the attack inflicts slow.
+---@field petrifies boolean Indicates whether the attack inflicts petrification.
+---@field plagues boolean Indicates whether the attack inflicts plague.
+---@field plague_type string If the attack inflicts plague, this specifies the unit type it is converted to.
+---@field backstabs boolean
+---@field rounds integer The number of rounds of combat.
+---@field firststrike boolean Indicates whether the attack always strikes first.
```

Comment:
I'm not sure, but suspect many of the booleans here are whether the ability is active in the particular fight (whether the opponent is immune to poison, also has firststrike, etc).